### PR TITLE
Fix zlib-ng build with cmake 3.31.0

### PR DESCRIPTION
**Pull request by am11**
We are seeing warnings like these (with cmake 3.31 in github actions):

```
CMake Deprecation Warning at D:/a/runtime/runtime/runtime/src/native/external/zlib-ng/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.10 will be removed from a future version of
  CMake.
```

When we started src/native/external projects, the idea was that we will keep our configs outside the vendor directory and we won't be using their cmake configs if they are providing one. That made version updates simple and "our" configs were just a flat list of files we want to build in the product. Now we are adding new dependencies with "fetch plan", that uses the cmake config from the vendor dir and we have to keep up with the config issues..

fyi @jkoritzinsky one downside of fetch plan.

### CODE REVIEW

**Comment by am11** [comment]
@jakobbotsch I'll revert the dummy change in JIT once windows format leg is green.

File: src/native/external/zlib-ng/CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,6 @@
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.5.1...3.31.0)
```
Comment (jakobbotsch) [inline src/native/external/zlib-ng/CMakeLists.txt:1]: Was this change meant to be in `cmake_minimum_required`? It seems odd to specify a range for the minimum required version.

Comment (am11) [reply src/native/external/zlib-ng/CMakeLists.txt:1]: The cmake folks just bumped the minimum version every after a while which breaks these kind of rigid contracts: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9875 (doesn't seem like they had a good reason to bump the version..).

As noted in PR description, the best way for us is to not use the FetchContent stuff and don't use their configs, as we have done for libunwind etc.

**Comment by am11** [comment]
@jakobbotsch, a side note: in https://github.com/dotnet/runtime/actions/runs/12145206891/job/33866308923?pr=110359 we can see repeated downloads:

```
Downloading 'https://clrjit2.blob.core.windows.net/clang-tools/17.0.6/windows-x64/clang-format.exe' to 'D:\a\runtime\runtime\runtime\artifacts\tools\clang-format.exe'
Downloading 'https://clrjit2.blob.core.windows.net/clang-tools/17.0.6/windows-x64/clang-tidy.exe' to 'D:\a\runtime\runtime\runtime\artifacts\tools\clang-tidy.exe'
```

first in `Build jitutils` step and then in `Run jitformat.py`. Perhaps we can skip the one in build step? Also, clang-tools can be bumped from v17 to v19.

File: src/coreclr/jit/abi.h [hunk 2]
```diff
@@ -1,6 +1,8 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+// dummy change
+
```
Comment (am11) [inline src/coreclr/jit/abi.h:5]:
```suggestion
```

**Comment by jakobbotsch** [comment]
I did not see the jit-format Windows run succeeding with this PR, it just seemed to be permanently stuck when I looked. Did you see it succeed?

**Comment by am11** [comment]
> I did not see the jit-format Windows run succeeding with this PR, it just seemed to be permanently stuck when I looked. Did you see it succeed?

The comment above it has a link to green run?

**Comment by MichalPetryka** [comment]
> I did not see the jit-format Windows run succeeding with this PR, it just seemed to be permanently stuck when I looked. Did you see it succeed?

Windows jit-format action has been broken on main for like 2 weeks, it's failing on every single JIT PR.

File: src/native/external/zlib-ng/CMakeLists.txt (hunk 1 above)
Comment (jkoritzinsky) [reply src/native/external/zlib-ng/CMakeLists.txt:1]: The CMake folks bump the minimum version to ensure that they can eventually remove support for old CMake policies.

Comment (jkoritzinsky) [reply src/native/external/zlib-ng/CMakeLists.txt:1]: Can we use `3.29.0` as the upper bound here to ensure we don't accidentally get new CMake policy choices we (and the zlib-ng build) aren't expecting?

```suggestion
cmake_minimum_required(VERSION 3.5.1...3.29.0)
```

We also will need to update `zlib-ng-version.txt` with this patch (and preferably upstream it).

Comment (am11) [reply src/native/external/zlib-ng/CMakeLists.txt:1]: @jkoritzinsky I see, well that makes sense. I was looking at their MR, it just pointed to previous similar version bump change.

Regarding "using vendor's cmake", we have our own conventions, best practices, version policies and update timelines. I think it's better to not rely on thirdparty configs and list the sources in our own scripts that we need to build for runtime use-cases. e.g. https://github.com/dotnet/runtime/blob/226ee0323375946d91cbac300bf83433539ae36c/src/native/external/llvm-libunwind.cmake#L7-L10

**Comment by jakobbotsch** [comment]
> The comment above it has a link to green run?

Ah, ok, looks good then. For some reason when I look at the "Checks" page it shows me a still running jit-format job.

> Windows jit-format action has been broken on main for like 2 weeks, it's failing on every single JIT PR.

Yes, @am11 is aiming to fix the failure with this PR.

**Comment by am11** [comment]
> Ah, ok, looks good then. For some reason when I look at the "Checks" page it shows me a still running jit-format job.

Yup, I had to update a file under coreclr/nativeaot and one eng/native to cleanup few other warnings. The problem is that the  python script fails when there is even a single config warning. Then GitHub running gets stuck for five hours in 'spawned process exiting with non-zero line' part of the pythong script. Maybe we should try to make the script robust against warnings.

**Comment by BruceForstall** [comment]
> first in Build jitutils step and then in Run jitformat.py. Perhaps we can skip the one in build step? Also, clang-tools can be bumped from v17 to v19.

It's fine to only download the tools once. It's not clear which download is actually getting used. I'm not familiar with the current job setup; @jkoritzinsky created that.

As for the version change, we shouldn't update unless we are required to. Or it provides large benefits of some kind. At one point there was an idea to share the tools between JIT and non-JIT runtime users, but I don't think that happened (it would make it much harder to update if it did).

**Comment by jkoritzinsky** [comment]
I specifically wanted to build `jit-utils` separately to avoid any problems with cloning and building jitutils within a runtime repo clone (which is what the bootstrap does). Building it within a runtime repo clone can cause issues with global.json and other configuration settings within the repo infra. Isolating it to its own sibling folder ensures that we don't have these issues.

We shouldn't be downloading jitutils again in the `Run jitformat.py` step. If we are, that's a bug.

**Comment by am11** [comment]
Now it's downloading once.

> I specifically wanted to build `jit-utils` separately to avoid any problems with cloning and building jitutils within a runtime repo clone (which is what the bootstrap does). Building it within a runtime repo clone can cause issues with global.json and other configuration settings within the repo infra. Isolating it to its own sibling folder ensures that we don't have these issues.

The python script clones it under $TEMP and does the "right thing". It won't clutter the diffs.

**Comment by jkoritzinsky** [comment]
You can remove the Checkout jitutils step as well then.

**Review by jakobbotsch — APPROVED** [review]
This looks good to me but I'm not very familiar with any of this, so I will defer to @jkoritzinsky for another approval.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Thanks!

**Pull request merged**